src/socket.rs

```diff
@@ -0,0 +1,335 @@
+use std::{
+    io,
+    mem::{self, MaybeUninit},
+    ptr,
+};
+
+pub struct Fd(usize);
+
+impl AsRawFd for Fd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.0 as RawFd
+    }
+}
+
+impl AsRawFd for &Fd {
+    fn as_raw_fd(&self) -> RawFd {
+        (*self).as_raw_fd()
+    }
+}
+
+impl Drop for Fd {
+    fn drop(&mut self) {
+        unsafe {
+            close(self).ok();
+        }
+    }
+}
+
+fn socket_addr(addr: SocketAddr) -> (*const sockaddr, socklen_t) {
+    match addr {
+        SocketAddr::V4(ref addr) => (
+            addr as *const _ as *const sockaddr,
+            size_of_val(addr) as socklen_t,
+        ),
+        SocketAddr::V6(ref addr) => (
+            addr as *const _ as *const sockaddr,
+            size_of_val(addr) as socklen_t,
+        ),
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+#[repr(i32)]
+pub enum SockType {
+    Stream = SOCK_STREAM,
+    Datagram = SOCK_DGRAM,
+    SeqPacket = SOCK_SEQPACKET,
+    Raw = SOCK_RAW,
+    Rdm = SOCK_RDM,
+}
+
+#[derive(Clone, Copy)]
+pub struct SockFlags(i32);
+
+impl SockFlags {
+    pub fn new() -> Self {
+        Self(0)
+    }
+
+    pub fn nonblock(mut self) -> Self {
+        self.0 |= SOCK_NONBLOCK;
+        self
+    }
+
+    pub fn cloexec(mut self) -> Self {
+        self.0 |= SOCK_CLOEXEC;
+        self
+    }
+}
+
+#[inline]
+pub fn socket(addr: SocketAddr, sock: SockType, flags: SockFlags) -> io::Result<Fd> {
+    // flags only supported by kernel >= 2.6.27
+    let domain = match addr {
+        SocketAddr::V4(_) => AF_INET,
+        SocketAddr::V6(_) => AF_INET6,
+    };
+    let ty = sock as i32 | flags.0;
+    // TODO expose protocol
+    let protocol = 0;
+    let res = unsafe {
+        syscall!(
+            Syscalls::Socket,
+            domain as isize,
+            ty as isize,
+            protocol as isize
+        )
+    };
+    result!(res).map(Fd)
+}
+
+#[inline]
+pub fn bind<F: AsRawFd>(socket: F, addr: SocketAddr) -> io::Result<()> {
+    let (addr, addr_len) = socket_addr(addr);
+    let res = unsafe {
+        syscall!(
+            Syscalls::Bind,
+            socket.as_raw_fd() as isize,
+            addr as isize,
+            addr_len as isize
+        )
+    };
+    result!(res).map(|_| ())
+}
+
+#[inline]
+pub fn getsockname<F: AsRawFd>(socket: F) -> io::Result<SocketAddr> {
+    let mut address = MaybeUninit::<libc::sockaddr_storage>::uninit();
+    let mut address_len = mem::size_of_val(&address);
+    let res = unsafe {
+        syscall!(
+            Syscalls::Getsockname,
+            socket.as_raw_fd() as isize,
+            address.as_mut_ptr() as isize,
+            &mut address_len as *mut _ as isize
+        )
+    };
+    let address = unsafe { address.assume_init() };
+    result!(res).map(|_| match address.ss_family as libc::c_int {
+        AF_INET => SocketAddr::V4(unsafe { ptr::read(&address as *const _ as _) }),
+        AF_INET6 => SocketAddr::V6(unsafe { ptr::read(&address as *const _ as _) }),
```

Comment:
Why are you casting a pointer to `libc::sockaddr_storage` to a pointer of a rust struct`SocketAddrV4/6`

Reply:
SocketAddrV4/V6 is just a wrapper around the libc::inet

Reply:
yeah, but a #[repr(Rust)] wrapper. it can have whatever size/alignment it wants.
there's no guarantee it was the same properties of what it wraps.